happi/backends/multi_db.py

```diff
@@ -0,0 +1,193 @@
+"""
+Backend implementation that combines multiple backends.
+"""
+import functools
+import logging
+import re
+from typing import Any, Dict, List, Optional, Union
+
+from .core import ItemMeta, ItemMetaGen, _Backend
+
+logger = logging.getLogger(__name__)
+
+
+def prevent_duplicate_ids(fn):
+    """
+    Decorator that remembers which document _id's have been yielded and
+    prevents subsequent documents with matching _id's from being yielded
+
+    Expects the wrapped function to yield ItemMeta documents
+    """
+    @functools.wraps(fn)
+    def wrapper(*args, **kwargs):
+        unique_ids = set()
+        for doc in fn(*args, **kwargs):
+            if doc['_id'] not in unique_ids:
+                unique_ids.add(doc['_id'])
+                yield doc
+
+    return wrapper
+
+
+class MultiBackend(_Backend):
+    """
+    Multi-backend backend.
+
+    Combines multiple backends, prioritizing database files in order
+    of appearance.
+
+    Parameters
+    ----------
+    backends : List[_Backend]
+        A list of instantiated backends in order of priority.
+        The backend at index 0 will have the highest priority.
+    """
+    def __init__(self, backends: List[_Backend]):
+        self.backends = backends
+
+    @property
+    def all_items(self) -> List[ItemMeta]:
+        """
+        List of all items in all backends.
+
+        In the case of duplicate items, the item in the first backend
+        takes priority.
+
+        Returns
+        -------
+        List[ItemMeta]
+            A list of all metadata documents
+        """
+        items = []
+        unique_ids = set()
+        for bknd in self.backends:
+            curr_items = bknd.all_items
+            for doc in curr_items:
+                if doc['_id'] not in unique_ids:
+                    items.append(doc)
+                    unique_ids.add(doc['_id'])
+
+        return items
+
+    def clear_cache(self) -> None:
+        """
+        Request to clear any cached data.
+
+        Clears the cache of each backend
+        """
+        for bknd in self.backends:
+            bknd.clear_cache()
+
+    @prevent_duplicate_ids
+    def find(self, *args, multiples=False, **kwargs) -> ItemMetaGen:
+        """
+        Find an instance or instances that matches the search criteria.
+
+        Parameters
+        ----------
+        multiples : bool
+            Find a single result or all results matching the provided
+            information.
```

Comment:
I think this keyword arg is actually unused here

Reply:
It's unused in all the other backends too.  I thought about implementing it here, but forgot to.  Is it something we care about?  I think we can just remove it for consistency.

Reply:
> It's unused in all the other backends too

Somehow I didn't notice, whoops! Maybe a cleanup problem for another time. It seems like the other backends just give multiples and don't define the kwarg at all...